src/Microsoft.DotNet.Interactive/RoutingSlipBase.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.DotNet.Interactive;
+
+public abstract class RoutingSlipBase
+{
+    private readonly List<Entry> _entries;
+
+    protected ICollection<Entry> Entries => _entries;
+
+    protected RoutingSlipBase(RoutingSlipBase source = null)
+    {
+        _entries = source switch
+        {
+            { } => new List<Entry>(source.ToUriArray().Select(e => new Entry { Uri = new Uri(e.AbsoluteUri), Completed = true })),
+            _ => new List<Entry>()
+        };
+    }
+    
+    protected class Entry
+    {
+        public Uri Uri { get; set; }
+        public bool Completed { get; set; }
+    }
+
+    public abstract void Stamp(Uri uri);
+
+    public Uri[] ToUriArray()
```

Comment:
I'm not sure this method should exist because it seems to lend itself to externalizing the core responsibilities of this class.